Create sampling response form

nathanArseneau opened the pull request:
<!-- Provide a brief summary of your changes -->
This address https://github.com/modelcontextprotocol/inspector/issues/199

## Motivation and Context
The sampling response was hard coded [here](https://github.com/modelcontextprotocol/inspector/blob/main/client/src/components/SamplingTab.tsx#L23), The goal was to remove the stubs.

## How Has This Been Tested?
I created a small MCP server to send / receive sampling request
![image](https://github.com/user-attachments/assets/529ad669-3553-4a0e-a395-81e98bb1f283)
![image](https://github.com/user-attachments/assets/bcb89961-0078-47e8-b9c0-30b157d4f89d)

![image](https://github.com/user-attachments/assets/32a2c782-d355-4e65-a405-624ec21b702a)

## Breaking Changes
Nope

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [X] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [X] My code follows the repository's style guidelines
- [X] New and existing tests pass locally
- [X] I have added appropriate error handling
- [X] I have added or updated documentation as needed

## Additional context
<!-- Add any other context, implementation notes, or design decisions -->

cliffhall commented:
Just want to leave this thought:

I think a JSON Editor should be an option here and not only the form. Two reasons: 
1. [A change](https://github.com/modelcontextprotocol/specification/pull/198) coming down the pike is that the `content` field can be an array and not just a single object. Supporting this will be complicated in the form. We'll need a mechanism for adding more content blocks, turning an object into the first element of an array and allowing editing of all of them.
2. We already have a similar problem with the Run Tool form, and this will be another form we have to figure that out for.

# Tool run form can't create new objects  (in this case states)
<img width="809" alt="Screenshot 2025-04-02 at 12 07 54 PM" src="https://github.com/user-attachments/assets/d1243710-dc93-4ffd-a7a2-e873b41bed1f" />

# Tool run form fortunately offers JSON editor option
<img width="797" alt="Screenshot 2025-04-02 at 12 07 38 PM" src="https://github.com/user-attachments/assets/e1caa422-4ae2-4608-9fa6-2cac61b2f4f9" />

# Tool run form doesn't know how to handle objects
<img width="778" alt="Screenshot 2025-04-02 at 12 07 30 PM" src="https://github.com/user-attachments/assets/3b8e1037-fc47-4146-a2c8-2d1465a8a45a" />

nathanArseneau commented:
<img width="1457" alt="image" src="https://github.com/user-attachments/assets/706e427a-5a0a-45bc-83c2-400745468d00" />
<img width="1535" alt="image" src="https://github.com/user-attachments/assets/1f7414da-173d-4407-b277-d7c9327c0109" />

cliffhall commented:
Hey @nathanArseneau. Testing this locally, I'm seeing something weird when trying to switch from JSON to the Form. The complaint comes when a hook in the `App.tsx` gets run again, trying to fetch config. No idea what has caused that hook to fail. The switch to form button is inside a component you're just using, but something must be different about the component where it is embedded. Can you have a look at how it is used in the ToolTab? It is not crashing there. This may not be about your implementation, but it's kind of a show stopper for this PR. 

### Switch to Form crashes in Sampling tab
https://github.com/user-attachments/assets/e73dc8c7-7c5e-43c1-abe4-d0bff2082169

### Switch to Form doesn't crash in Tools tab
https://github.com/user-attachments/assets/49f09836-259e-4dca-b55c-3c1729eccf14

nathanArseneau commented:
Sure, I will be happy to investigate, and yes, it is better we solve this issue before this PR

however, when I am testing from my end, I am not able to reproduce

nathanArseneau commented:
My apologies; that was my mistake. The sampling is wrapped in a form for semantics, and the buttons were not typed.

I did not get your error with the config, but I managed to reproduce a similar behavior.

olaservo commented on client/src/components/SamplingRequest.tsx:
```diff
@@ -0,0 +1,162 @@
+      properties: {
+        model: {
+          type: "string",
+          default: "GPT-4o",
```
Since this isn't really sampling from a real LLM anyway (the result is being edited by hand), does this need to default to the name of a real model?

nathanArseneau replied:
We could use the same value as before, "stub-model." However, we definitely need a value for this field.

Has an increment on this. I was thinking that we could save those responses and have the ability to select from a dropdown list to resend past sampling responses. This way, the model name could be defined once and saved based on the user's preference.

olaservo replied:
Lets have the default as `stub-model` just to go with the spirit of this not being specific to a model provider.

Re: your improvement idea, that could make sense - lets save that for a separate PR though?

cliffhall reviewed (changes requested):
Runs locally and does the right stuff now. However eslint is reporting some errors on the new `SamplingRequest.tsx` file. 

```shell

cd client
eslint


Projects/mcp-inspector/client/src/components/SamplingRequest.tsx
   69:43  error    Unexpected any. Specify a different type                                                                                                @typescript-eslint/no-explicit-any
  111:6   warning  React Hook useMemo has a missing dependency: 'messageResult'. Either include it or remove the dependency array                          react-hooks/exhaustive-deps
  111:7   warning  React Hook useMemo has a complex expression in the dependency array. Extract it to a separate variable so it can be statically checked  react-hooks/exhaustive-deps
  111:25  error    Unexpected any. Specify a different type                                                                                                @typescript-eslint/no-explicit-any

✖ 4 problems (2 errors, 2 warnings)

```

nathanArseneau commented:
Fixed and added in CI to avoid having to run locally to validate.
![image](https://github.com/user-attachments/assets/6990c29c-59e1-40ce-84fc-0ce7cfc68c9d)

build will fail with any
![image](https://github.com/user-attachments/assets/a9b0832f-2023-4014-bce2-c57aa8454b07)

cliffhall commented on client/src/components/SamplingRequest.tsx:
```diff
@@ -0,0 +1,168 @@
+      properties: {
+        model: {
+          type: "string",
+          default: "GPT-4o",
```
I thought we were going with `stub-model` here?

cliffhall reviewed (changes requested):
Looks good and runs. Still have a specific model name in the component though.

cliffhall commented:
@nathanArseneau A wrinkle here is that a [recent change](https://github.com/modelcontextprotocol/inspector/pull/284) to the `DynamicJsonForm` has made it so that we do not attempt to display complex forms. If the schema to be edited is anything but a list of basic fields, the JSON form is the only affordance. So this only displays the JSON form now. See that PR for details on that decision.

Also, I notice that when I approve my edits, the response sent back to the server is not complete.

### The [response format](https://modelcontextprotocol.io/docs/concepts/sampling#response-format):
<img width="619" alt="Screenshot 2025-04-23 at 4 45 27 PM" src="https://github.com/user-attachments/assets/df846970-c355-481f-9f9d-b19c3e8ede6a" />

### Response actually sent from the client
<img width="791" alt="Screenshot 2025-04-23 at 4 43 53 PM" src="https://github.com/user-attachments/assets/15be4484-e6e1-4da8-8137-16ef56c7dc85" />

nathanArseneau commented:
While fixing the merge conflicts, I saw that change and personally was not a fan of it. 

[The fix ](https://github.com/modelcontextprotocol/inspector/pull/284/files#diff-0679048afea6c4dc4e0aada89c208d368d7bff3f3c9e4502e1a982cb2a0e1b96R16) was to remove the [Object: object], but in my opinion, the component should have been extended to support more complex schemas, instead of removing the form entirely when the schema is 'complex' or not flat.  Another issue should be opened to fix this behavior.

So I assume your test scenario is: 
You have a tool that creates a sampling request and waits for a response. The response from your tool call is the response for the sampling.

![image](https://github.com/user-attachments/assets/60ef49d0-6df6-4380-83db-9d3c60edc415)

However, [sampling response type](https://github.dev/modelcontextprotocol/typescript-sdk/blob/131776764536b5fdca642df51230a3746fb4ade0/src/types.ts#L1022) and [tool response type](https://github.dev/modelcontextprotocol/typescript-sdk/blob/131776764536b5fdca642df51230a3746fb4ade0/src/types.ts#L867-L868) are different, which is why you do not see the expected outcome.  If I'm wrong in my assumption, I'd be happy to implement a fix if you provide your test scenario.

Here is the code for my tool
```ts
server.tool(
    'sampling',
    'This is a tool to test sampling',
    { str: z.string() },
    async ({ str }) => {
        const samplingResponse = await server.server.createMessage({
            messages: [
                {
                    role: 'user',
                    content: {
                        type: 'text',
                        text: `sampling request for ${str}?`,
                    },
                },
            ],
            systemPrompt: 'You are a helpful file system assistant.',
            includeContext: 'thisServer',
            maxTokens: 100,
        })

        return {
            content: [
                {
                    type: 'text',
                    text: JSON.stringify(samplingResponse, null, 2),
                },
            ],
        }
    }
)
```

cliffhall commented:
> @cliffhall While fixing the merge conflicts, I saw that change and personally was not a fan of it.
> in my opinion, the component should have been extended to support more complex schemas, instead of removing the form entirely when the schema is 'complex' or not flat. Another issue should be opened to fix this behavior.

I get it. The problem is schemas of arbitrary depth are going to result in a very complex form. Our existing code for creating the forms was not optimal, and ad-hoc extending that to handle _anything_ seemed unwise. The form would be nice, but not in the form of a maintenance burden. Definitely another issue/PR that tackles it in isolation is the way to go.

cliffhall commented:
> However, [sampling response type](https://github.dev/modelcontextprotocol/typescript-sdk/blob/131776764536b5fdca642df51230a3746fb4ade0/src/types.ts#L1022) and [tool response type](https://github.dev/modelcontextprotocol/typescript-sdk/blob/131776764536b5fdca642df51230a3746fb4ade0/src/types.ts#L867-L868) are different, which is why you do not see the expected outcome.

Yep, you're right. That's the server echoing back the text from the sampling response. Not an issue.

cliffhall reviewed (approved):
Although we need to address dynamic form building in another request, this PR gives us the ability to edit the previously stubbed LLM response that goes back to the server, which will be useful for testing the functionality of server tools that request sampling.

<img width="1584" alt="edit sampling response" src="https://github.com/user-attachments/assets/364d6ab4-8751-4bfe-835e-8af6443dd2e5" />

cliffhall commented:
This fixed #199

MAX96811 commented:
WOW!
